File: csharp/simple-cipher/Cipher.cs
namespace simple_cipher
{
    using System;
    using System.Linq;

    public class Cipher
    {
        private const string Letters = "abcdefghijklmnopqrstuvwxyz";
        private const int KeyLength = 100;

        private static readonly Random Random = new Random();

        public Cipher()
            : this(GenerateKey())
        {
        }

        public Cipher(string key)
        {
            if (IsInvalidKey(key))
            {
                throw new ArgumentException("Invalid key specified.", "key");
            }

            this.Key = key;
        }
        
        public string Key { get; private set; }

        public string Encode(string plainText)
        {
            return new string(plainText.Select(this.Encode).ToArray());
        }

        private char Encode(char c, int i)
        {
            return this.Shift(c, i, (character, shift) => character + shift);
        }

        public string Decode(string encryptedText)
        {
            return new string(encryptedText.Select(this.Decode).ToArray());
        }

        private char Decode(char c, int i)
        {
            return this.Shift(c, i, (character, shift) => character - shift);
        }

        private char Shift(char c, int i, Func<int, int, int> shiftCharacter)
        {
            return (char)('a' + this.GetCharacterShiftValue(c, i, shiftCharacter));
        }

        private int GetCharacterShiftValue(char c, int i, Func<int, int, int> shiftCharacter)
        {
            return Mod(shiftCharacter(NormalizeCharacter(c), this.GetShiftValue(i)), Letters.Length);
        }

        private static int NormalizeCharacter(char c)
        {
            return c - 'a';
        }

        private int GetShiftValue(int i)
        {
            return this.Key[i] % 'a';
        }

        private static int Mod(int x, int y)
        {
            return ((x % y) + y) % y;
        }

        private static string GenerateKey()
        {
            return new string(Enumerable.Repeat(Letters, KeyLength)
                .Select(s => s[Random.Next(s.Length)])
                .ToArray());
        }

        private static bool IsInvalidKey(string key)
        {
            return string.IsNullOrEmpty(key) || !key.All(c => Letters.Contains(c));
        }
    }
}

File: java/simple-cipher/src/main/java/Cipher.java
import java.util.Random;
import java.util.function.BiFunction;

public class Cipher {
    private static final String LETTERS = "abcdefghijklmnopqrstuvwxyz";
    private static final int KEY_LENGTH = 100;

    private static final Random random = new Random();

    private final String key;

    public Cipher() {
        this.key = generateKey();
    }

    public Cipher(String key) {
        if (!key.matches("^[a-z]+$")) {
            throw new IllegalArgumentException();
        }

        this.key = key;
    }

    public String getKey() {
        return key;
    }

    public String encode(String input) {
        String encoded = "";

        for (int i = 0; i < input.length(); i++) {
            encoded += encode(input.charAt(i), i);
        }

        return encoded;
    }

    public String decode(String encoded) {
        String decoded = "";

        for (int i = 0; i < encoded.length(); i++) {
            decoded += decode(encoded.charAt(i), i);
        }

        return decoded;
    }

    private char encode(char c, int i) {
        return this.shift(c, i, (character, shift) -> character + shift);
    }

    private char decode(char c, int i) {
        return this.shift(c, i, (character, shift) -> character - shift);
    }

    private char shift(char c, int i, BiFunction<Integer, Integer, Integer> shiftCharacter) {
        return (char) ('a' + this.getCharacterShiftValue(c, i, shiftCharacter));
    }

    private int getCharacterShiftValue(char c, int i, BiFunction<Integer, Integer, Integer> shiftCharacter) {
        return mod(shiftCharacter.apply(normalizeCharacter(c), this.getShiftValue(i)), LETTERS.length());
    }

    private static int normalizeCharacter(int c) {
        return c - 'a';
    }

    private int getShiftValue(int i) {
        return this.key.charAt(i) % 'a';
    }

    private static int mod(int x, int y) {
        return ((x % y) + y) % y;
    }

    private static String generateKey() {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < KEY_LENGTH; i++) {
            int index = random.nextInt(26);
            sb.append(LETTERS.charAt(index));
        }

        return sb.toString();
    }
}